### ISSUE TITLE
rest parameter ts2556 and ts2345

### ISSUE THREAD

**Issue by liukonghe** [microsoft/TypeScript#57170]
### 🔎 Search Terms

 ts2556 ts2345

### 🕗 Version & Regression Information

version: 5.3.3

### 💻 Code

```ts
const BoxGeometry = (_width: number, _height: number) => undefined;
const SphereGeometry = (_radius: number, _height: number, _openEnded?: boolean) => undefined;

export const Geometry = {
  Box: BoxGeometry,
  Sphere: SphereGeometry,
};

type GeometryType = keyof typeof Geometry;

export type GetArgument<T extends GeometryType> = Parameters<(typeof Geometry)[T]>;

export function load<T extends GeometryType>(type: T, args: GetArgument<T>) {
  const geometry = Geometry[type](...args);
}
```

### 🙁 Actual behavior

![image](https://github.com/microsoft/TypeScript/assets/18611375/f39a312f-7fd4-4b1c-9f4d-7643570ff540)

### 🙂 Expected behavior

...args should not report an error

**Comment by Andarist** [comment]
Likely this particular error would get fixed with https://github.com/microsoft/TypeScript/pull/50034 but it wouldn't fix the correlation problem. To fix the latter (and have it working today) you can use the pattern introduced by TS 4.6 ([here](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-6.html#indexed-access-inference-improvements)):
```ts
const BoxGeometry = (_width: number, _height: number) => undefined;
const SphereGeometry = (
  _radius: number,
  _height: number,
  _openEnded?: boolean,
) => undefined;

type TypeMap = {
  Box: Parameters<typeof BoxGeometry>;
  Sphere: Parameters<typeof SphereGeometry>;
};

type GeometryType<P extends keyof TypeMap = keyof TypeMap> = {
  [K in P]: (...args: TypeMap[K]) => undefined;
};

export const Geometry: GeometryType = {
  Box: BoxGeometry,
  Sphere: SphereGeometry,
};

export function load<K extends keyof TypeMap>(type: K, args: TypeMap[K]) {
  const geometry = Geometry[type](...args);
}
```

**Comment by craigphicks** [comment]
Isn't there some overthinking going on here?  Is the following not an example realistic scenario? If more types are added `load2` might have to do more type checking.


```ts
interface Geometry2 {
  Box: (width:number,height:number) => undefined,
  Sphere: (radius:number,height:number,openEnded?: boolean) => undefined,
  Cube: (width:number,length:number, height:number) => undefined,
};

declare const g: Geometry2;

// key and args set from user form
declare const key: keyof Geometry2;
declare const args: [wr:number,h:number,o?:boolean|number];

function load2(key: keyof Geometry2, arg0: number, arg1: number, arg2?: boolean|number) {
    switch (key) {
        case "Box": return g.Box(args[0],args[1]);
        case "Sphere": {
            if (typeof arg2 !== "boolean") throw "rangeError";
            return g.Sphere(args[0],args[1],args[2]);
        }
        case "Sphere": {
            if (typeof arg2 !== "number") throw "rangeError";
            return g.Sphere(args[0],args[1],args[2]);
        }
    }
    throw "rangeError";
}
try {
    load2(key,...args)
} catch (e){}
```
Or, if you are really certain that inputs will never be out of bounds, or you are going to call the `load` only with literals (seems unlikely):
```ts
function load3<K extends keyof Geometry2>(key: K, args: Parameters<Geometry2[K]>) {
    // @ts-ignore
    g[key](...args);
}
load3("Box",[1,1]);
load3("Sphere",[1,1,true]);
load3("Box",[1]); // error
load3("Sphere",[1,1,1]); // error
```

**Comment by RyanCavanaugh** [comment]
The example + links in https://github.com/microsoft/TypeScript/issues/57170#issuecomment-1909905713 is the best we can do without potentially combinatorial explosion

**Comment by craigphicks** [comment]
Fundamentally the same, but it can encapsulated in a simple reusable type function `RemapWorkaround`, that is easy to apply, and can applied locally at the point of error:

```ts
type RemapWorkaround<G extends Record<keyof G,(...args:any[])=>any>, P extends keyof G = keyof G> = {
  [K in P]: (...args: Parameters<G[K]>) => ReturnType<G[K]>;
}

interface Geometry {
    Box: (width:number,height:number) => 1,
    Sphere: (radius:number,height:number,openEnded?: boolean) => 2,
    Cube: (width:number,length:number, height:number) => 3,
};

declare const g: Geometry;

function load<K extends keyof Geometry>(key: K, ...args: Parameters<Geometry[K]>): ReturnType<Geometry[K]> {
    args satisfies Parameters<Geometry[K]>; // no error (as expected)
//    const r = g[key](...args); // error (limitation)
//                   ~~~~~~~
//A spread argument must either have a tuple type or be passed to a rest parameter.(2556)

    // Use RemapWorkaround as an inline fix
    const r = (g as RemapWorkaround<Geometry>)[key](...args); // no error
    return r;
}
const r1 = load("Box",1,1);
const r2 = load("Sphere",1,1,true);
const r3 = load("Cube",1,1,1);

r1 satisfies ReturnType<Geometry["Box"]>
r2 satisfies ReturnType<Geometry["Sphere"]> 
r3 satisfies ReturnType<Geometry["Cube"]>

load("Box",1); // error (as expected)
load("Cube", 1,1,1,1); // error (as expected)
load("Cube", 1,1,1n); // error (expected error because last one is not a number)

load("Sphere", 1,1,1); // error (as expected)
load("Sphere", 1,1,undefined); // error (as expected)
load("Cube", 1,1,true); // error (as expected)
```
[playground](https://www.typescriptlang.org/play?ts=5.1.6#code/C4TwDgpgBAShC2BDMB1A9gJwNaI2grgHYAmAPAOJQQAewEJAzrBAMaZlYQhoBmU5AGgAUAOjG4A5gwBciQiADaAXQCUAXgB8ckBoFQAClVr1iTTtz6U1Uc734ao1gN4BYAFBQoCgNJQAloQGStJQouIYUiH6uIjwEHQYDBQ+Shoqjg5wwPgYhAAq4BDJ3qkA3O4Avu7uAQk8iCzQ5BBoccAYIFCuHp5QAEJo1CFCAO5+xMAAFtKE+PAARhAYApMQfhKTwDNzixjpmlAAjALuvVAAymCrGBDDGIjEfvgyswtLK2sbW6+7AmiQhAAoiQIMQAPwheZoNAAGwgcn2DgATCcep4AML4RbDMYTaY-d5wwgSKbbN7LKCrdabMm7RFQADMqIq5Tc7mIrBhuGgbEIDGAUAkIWarXiHVZ7h4RBYwD8aECMLQD1IvhodEYNi4dhFbQ6GiE5hC3j0YhEkhkBhibSWSR1YsUJTSISyOXyhQoLV1DtSXVOvXNUAYiFlDB4fggTGi92tiQ9ova3o0pSgAHoU1BCGgqBg8BhQogmDRIDLQSp3Gmzrz+VA89YJApzEowmaIgwVMmK0tc6EYX54H5gMG5YQy24K2cJ5PegA-Wdz6fllMAQUDYBuDygkjm9AF8GeArWUyWlMQADdoIgoNkwHCr4UoJgoIsoGACwxQVes5ebtXX9H4ksIhCEiACsIEAGyjn6ngVgAqu+zBIKgmA4HgRDEJuTByP4hC9oQ0BhtQ0FQFWAq1qEEiYYhyDoNguAECQcZemkDZcE2prmu2qbppm2a5sRNzZLkNaslUbikTWhyOFAioPEIABEAzUPJAjHIc7buBJGBItJsnEAplzXBAKlqQI7T4BAGnifK1YYAyulKvp8mYosJmqapVnuBgUlBiGYYRswQlupATH2goimDPJqReTpvl+KG4ZMC6uQFCFdoJuFhlLMZPpefZcUJQFyXBUU6UdOFLk5Ro1RuHpClKSZXGdjmj5CAWRjFnQxCjnVzlYsZeimWpTXpl2rXtUWrBdT1jkKZVKlHO5hwjh2o0tXmQiTSWGFjXmiwsIgzzQFy1bytA8UZmgAqXgSew1b1WU3AtQ0jXx42FtQnWlu4D1XNlz3uehEBhvh3WrW9G0TZ9U3fbVs19a5g3ueZlng7t+YfV93XuEAA)

**Comment by Andarist** [comment]
So it seems that the change from https://github.com/microsoft/TypeScript/pull/50034 (merged as part of https://github.com/microsoft/TypeScript/pull/57122 ) didn't fix the "A spread argument must either have a tuple type or be passed to a rest parameter.(2556)" problem here. I think that part should still work, since TS is able to verify that `...args` are spreadable in other circumstances ([TS playground](https://www.typescriptlang.org/play?ts=5.5.0-dev.20240227#code/MYewdgzgLgBAQiAHgcQKYgLaqgJwJ4wC8MAFAPoDuAlgCZQAWAXDGAK4YBGqOANDGfVRUA5vSjM2nbgEoiAPhiswNVADMqYVDQDcAKFCRYAZQAOgnKjSZs+IqV0x+OAIY0qrCBPZdeD-oJExLylfRzIQE1QwAFFlLQB+Zg4QEAAbVGcwHl1ZQgUlFXVNHV1dVEQTEBxYA2gYKyxcAmIAbz8ERGYOhps8bMdTc1RmQe5LdEb8bIBfPV0oPEj6id6AFUXUOwBrVDwQVRgFyP3l6ya58srqw43lqABBHGF2KKgAHlWYcqgomghTyZ4daRBTEAAKzhcjW4EDefhIR3QBx6TWkAG1VgBdXRyOaqJTAKBUcAwZzAYCoEwPKAAGQy0AA8poSM5mJk+jAAHTcyHCTyksB4NGY2QtaalS5VWD4sCE4lgGCpECuD5fRA-ZT-FH4YGoOQIjbMVZ8Xn8tAPJ4vMDvVZyUV+WqwYQrJp2bVCxGYkjczmm6R6RxkilU+60+lQJmob08p4Qf2lRyOtUalQ0R58gCMdjRYr4PtNmNJ-0dAZgSe+vy06YgACZmBZXOBUgQ0ey8zzBcLC8Qc9N277Y4XnMXwNA9OKgA)):

```ts
const BoxGeometry = (_width: number, _height: number) => undefined;
const SphereGeometry = (
  _radius: number,
  _height: number,
  _openEnded?: boolean,
) => undefined;

export const Geometry = {
  Box: BoxGeometry,
  Sphere: SphereGeometry,
};

type GeometryType = keyof typeof Geometry;

export type GetArgument<T extends GeometryType> = Parameters<
  (typeof Geometry)[T]
>;

function acceptAtLeastOne(a: any, ...args: any[]) {}

export function load<T extends GeometryType>(type: T, args: GetArgument<T>) {
  const geometry = Geometry[type](...args); // error (but a wrong one)
  acceptAtLeastOne(...args); // error

  const extendedArgs1 = [{}, ...args] as const; // ok
  const extendedArgs2: readonly [any, ...any[]] = [{}, ...args] as const; // ok
}
```
